jqjq.jq

```diff
@@ -2456,35 +2478,85 @@ def fromjqtest:
-      else {filter: .[0]}, (.[1:] | _f)
+      elif $a == "-h" or $a == "--help"              then {help: true}, (.[1:] | _f)
+      elif $a == "--jq"                              then {jq: .[1]}, (.[2:] | _f)
+      elif $a == "--lex"                             then {lex: true}, (.[1:] | _f)
+      elif $a == "--no-builtins"                     then {no_builtins: true}, (.[1:] | _f)
+      elif $a == "--parse"                           then {parse: true}, (.[1:] | _f)
+      elif $a == "--repl"                            then {repl: true}, (.[1:] | _f)
+      elif $a == "-n" or $a == "--null-input"        then {null_input: true}, (.[1:] | _f)
+      elif $a == "-c" or $a == "--compact-output"    then {compact_output: true}, (.[1:] | _f)
+      elif $a == "-r" or $a == "--raw-output"        then {raw_output: true}, (.[1:] | _f)
+      elif $a == "--raw-output0"                     then {raw_output: true,
+                                                           raw_no_lf: true,
+                                                           raw_output0: true}, (.[1:] | _f)
+      elif $a == "-j" or $a == "--join-output"       then {raw_output: true,
+                                                           raw_no_lf: true}, (.[1:] | _f)
+      elif $a == "-s" or $a == "--slurp"             then {slurp: true}, (.[1:] | _f)
+      elif $a == "-C" or $a == "--color-output"      then {color_output: true}, (.[1:] | _f)
+      elif $a == "-M" or $a == "--monochrome-output" then {monochrome_output: true}, (.[1:] | _f)
+      elif $a == "--run-tests"                       then {run_tests: true}, (.[1:] | _f)
+      elif $a == "--"                                then {filter: .[1]}, (.[2:] | _f)
+      elif $a | startswith("-")                      then error("unknown argument: \($a)")
+      else {filter: $a}, (.[1:] | _f)
       end;
     ( [_f]
     | add
     );
 
+  # get the ANSI color codes for printing values
+  # corresponds to jv_set_colors in jq and its usage in main
+  def _parse_colors($opts; $env):
+    # color order: null, false, true, number, string, array, object, field
+    ( ["0;90", "0;39", "0;39", "0;39", "0;32", "1;39", "1;39", "1;34"] as $default
+    | if $env | has("JQ_COLORS") then
+        # only up to the first 8 color sequences are used
+        ( ($env.JQ_COLORS | split(":")[:8]) as $custom
+        # jq limits color sequences to 12 bytes, to fit into 16-byte buffers
+        # with ESC, [, m, and NUL.
+        | if $custom | all(length <= 12 and test("^[0-9;]*$")) then
+            $custom + $default[$custom | length:]
+          else
+            "Failed to set $JQ_COLORS\n" | stderr | $default
+          end
+        )
+      else $default
+      end
+    | if $opts.monochrome_output or
+          (($opts.color_output | not) and ($env.NO_COLOR | . != null and . != "")) then
+        null
+      end
+    );
+
+  def _parse_opts($opts; $env):
```

Comment:
I think `$env` is available here (and for `_parse_colors `) as a argument to `_jqjq` but maybe it's confusing?

Reply:
Oh I see that now. I'm tempted to leave it there, to make the dependency explicit, but up to you.

Reply:
Got confused too :) maybe leave as is but rename either this arg or both to be clearer somehow?

Reply:
I'd rather leave it for now. When the option parsing is split out into a separate jq module to move logic from bash, then I think it'll be more clear how it should finally end up.